sycl/doc/extensions/Avoid_event_creation/avoid_event_creation.asciidoc

```diff
@@ -0,0 +1,144 @@
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+
+== Notice
+
+IMPORTANT: This specification is a draft.
+
+Copyright (c) 2021 Intel Corporation. All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.
+GitHub does not render image icons.
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Introduction
+
+This extension adds `avoid_event_creation` property for `sycl::queue`, by using this property
+the application informs SYCL RT library that it doesn't care about the returned event,
+i.e the application doesn't use the returned event and just ignores it.
+
+In that case, the RT library makes a decision about internal optimization - voiding
+the creation of low-level event from the backend. The property affects only kernel cases,
+i.e. avoid creating low-level event for submission of kernels.
+Performance improvement is the main motivation for using this property.
+
+Below is a usage example:
+
+[source,c++]
+----
+  // in order queue is required for avoid_event_creation
+  sycl::queue Queue( {property::queue::in_order{}, property::queue::avoid_event_creation{}} );
+
+  // some USM preparations ..
+
+  // creating "null" events
+  sycl::event e1, e2, e3;
+
+  // returning "null" events from each submission function:
+  e1 = Queue.parallel_for(NDRange, [=](nd_item<1> item){ do_smth1(); });
+
+  e2 = Queue.single_task([=](){ do_smth2(); });
+
+  e3 = Queue.submit([&](handler &CGH) { CGH.parallel_for(NDRange, [=](nd_item<1> item){ do_smth3(); }); });
+
+  Queue.wait();
+----
+
+In the example above, the application doesn't use sycl events: `e1`, `e2`, `e3`
+and is waiting for the end of work by `queue::wait()`. The returned events will be
+`null events` - means that low-level events from backend are not created for them.
+`null event` is the event created by the default event constructor, i.e `null event`
+has the same behavior as if the event was created by the default constructor.

```

Comment:
I think the behavior here is subtle and potentially dangerous.

If you keep this behavior, then I think you need to expand on the discussion here with examples and non-normative notes to help users reason about what will happen.

SYCL 2020 says that the default constructor:

> Constructs an event that is immediately ready. The event has no dependencies and no associated commands. Waiting on this event will return immediately and querying its status will return `info::event_command_status::complete`.

If I'm reading your proposal and SYCL 2020 correctly, that means the following code examples won't give compilation or run-time errors, but also won't do what the user expects:
```c++
sycl::event e1;
e1 = q.parallel_for(...);
e1.wait(); // Returns immediately (because e1 is in "complete" status) and does not wait for the kernel
```

```c++
sycl::event e1, e2;
e1 = q1.parallel_for(...);
e2 = q2.parallel_for(e1, ...); // Runs immediately (because e1 is in "complete" status)
e2.wait();
```

Have you considered whether it would be possible to create a special event, instead, or perhaps a new `info::event::command_execution_status`?  If the event returned by one of these special queues had its status set not to `complete` but to something like `invalid` instead, we could also throw an `errc::invalid_operation` if somebody tried to wait on the event or pass it as an argument to some sort of dependency-expressing function.

Reply:
@Pennycook, I originally thought it was a good idea to return the default event when we have the new property is set. But for cases like multiple queues seems it makes sense to have different behaviour for that. For wait() probably also. I updated the document.